python/vineyard/core/client.py

```diff
@@ -0,0 +1,422 @@
+from vineyard._C import BlobBuilder
+from vineyard._C import Object
+from vineyard._C import ObjectID
+from vineyard._C import ObjectMeta
+from vineyard._C import RemoteBlob
+from vineyard._C import RemoteBlobBuilder
+from vineyard._C import _connect
+
+
+class Client:
+    """
+    Client is responsible for managing IPC and RPC clients for Vineyard
+    and provides a high-level interface to fetch objects from the Vineyard cluster.
+    """
+
+    def __init__(self):
+        # Initialize ipc and rpc clients based on the type of the given client
+        self._ipc_client = None
+        self._rpc_client = None
+
+    def connect(self, *args, **kwargs) -> 'Client':
+        """
+        Connects to the vineyard IPC socket and RPC socket.
+
+        The function supports connecting by the following priority:
+        - Using an explicit IPC socket or RPC endpoint.
+        - Using IPC socket or RPC endpoint from environment variables
+          VINEYARD_IPC_SOCKET or VINEYARD_RPC_ENDPOINT.
+
+        If providing an explicit IPC socket, and the IPC socket environment
+        variable is also set, the explicit IPC socket will be used. Also, if
+        the RPC endpoint environment variable is set at the same time, the
+        RPC endpoint will be used as well.
+        """
+
+        try:
+            client = _connect(*args, **kwargs)
+        except Exception:
+            client = None
+        if client:
+            if client.is_ipc:
+                self._ipc_client = client
+            else:
+                self._rpc_client = client
+
+        # Attempt to connect using environment variables
+        username = kwargs.get('username', "")
+        password = kwargs.get('password', "")
+        session_id = kwargs.get('session_id', 0)
+
+        if self._ipc_client is None:
+            ipc_socket = os.getenv('VINEYARD_IPC_SOCKET')
+            if ipc_socket and os.path.exists(ipc_socket):
+                # Connect using IPC socket
+                self._ipc_client = _connect(
+                    ipc_socket, username=username, password=password
+                )
+
+        if self._rpc_client is None:
+            rpc_endpoint = os.getenv('VINEYARD_RPC_ENDPOINT', None)
+            if rpc_endpoint:
+                # resolve rpc endpoint from environment variable
+                hostname, port_str = rpc_endpoint.split(':')
+                port = int(port_str)
+                self._rpc_client = _connect(
+                    hostname,
+                    port,
+                    session_id=session_id,
+                    username=username,
+                    password=password,
+                )
+
+        if self._ipc_client is None and self._rpc_client is None:
+            raise ConnectionError(
+                "Failed to resolve IPC socket or RPC endpoint of vineyard server from "
+                "environment variables VINEYARD_IPC_SOCKET or VINEYARD_RPC_ENDPOINT."
+            )
+        return self
```

Comment:
Please put contents in `connect()` to `__init__()` to make it more Pythonic.

Reply:
Done.